File: spk/_global.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import Union

from . import api, storage


def load_spec(pkg: Union[str, api.Ident]) -> api.Spec:
    """Load a package spec from the default repository."""

    if not isinstance(pkg, api.Ident):
        pkg = api.parse_ident(pkg)

    try:
        return storage.remote_repository().read_spec(pkg)
    except storage.PackageNotFoundError:
        return storage.local_repository().read_spec(pkg)


def save_spec(spec: api.Spec) -> None:
    """Save a package spec to the local repository."""

    repo = storage.local_repository()
    repo.force_publish_spec(spec)

File: src/global.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use std::{convert::TryInto, sync::Arc};

use crate::{
    api,
    storage::{self, Repository},
    Error, Result,
};

#[cfg(test)]
#[path = "./global_test.rs"]
mod global_test;

/// Load a package spec from the default repository.
pub async fn load_spec<S: TryInto<api::Ident, Error = crate::Error>>(
    pkg: S,
) -> Result<Arc<api::Spec>> {
    let pkg = pkg.try_into()?;

    // Do not require "origin" to exist.
    match storage::remote_repository("origin").await {
        Ok(repo) => match repo.read_spec(&pkg).await {
            Ok(spec) => return Ok(spec),
            Err(Error::PackageNotFoundError(_)) => {}
            Err(err) => return Err(err),
        },
        Err(Error::SPFS(spfs::Error::FailedToOpenRepository { source, .. }))
            if matches!(*source, spfs::Error::UnknownRemoteName(_)) => {}
        Err(err) => return Err(err),
    }

    storage::local_repository().await?.read_spec(&pkg).await
}

/// Save a package spec to the local repository.
pub async fn save_spec(spec: &api::Spec) -> Result<()> {
    let repo = storage::local_repository().await?;
    repo.force_publish_spec(spec).await
}
